File: proof_kernel/hypothesis_registry.py
"""Baseline hypothesis registry.

Each experiment in the suite must declare the baseline hypothesis it is testing.
The registry is the single source of truth for:

- what hypothesis an experiment is testing,
- what its expected signature is,
- what its failure does and does not mean,
- what alternative hypotheses are admissible,
- what candidate lemma the experiment may suggest.

The registry is split across role-based files in proof_kernel/hypotheses/. The
canonical baseline is read-only here; agent-proposed revisions go through the
hypothesis_proposals workflow and never silently mutate this registry.

See PROOF_PROGRAM_SPEC.md for the rationale.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Dict, Iterable, List, Optional

SCHEMA_VERSION = "2026.05.hypothesis-registry.v1"

REGISTRY_FILES = (
    "program_1.json",
    "program_2.json",
    "controls.json",
    "pathfinders.json",
    "demonstrations.json",
)

REQUIRED_EXPERIMENT_IDS = (
    "EXP_0",
    "EXP_1",
    "EXP_1B",
    "EXP_1C",
    "EXP_2",
    "EXP_2B",
    "EXP_3",
    "EXP_4",
    "EXP_5",
    "EXP_6",
    "EXP_7",
    "EXP_8",
    "EXP_9",
    "EXP_10",
)

VALID_ROLES = {"witness", "control", "pathfinder", "demonstration", "exploratory", "visualization"}
VALID_PROGRAMS = {"PROGRAM_1", "PROGRAM_2", "NONE"}


def _registry_dir(repo: Path | str = ".") -> Path:
    return Path(repo) / "proof_kernel" / "hypotheses"


def load_registry(repo: Path | str = ".", apply_overlays: bool = True) -> Dict[str, Any]:
    """Load and merge the baseline hypothesis registry.

    By default applies any accepted hypothesis proposals on top of the canonical
    JSON files. Pass apply_overlays=False to read the canonical-only registry
    (used by the proposal workflow itself, to compare current vs proposed).

    Returns a dict with:
      - by_hypothesis_id: { hypothesis_id -> entry }
      - by_experiment_id: { experiment_id -> entry }
      - by_display_id:    { display_id -> entry }
      - all:              [ entry, ... ]
      - sources:          [ filename, ... ]
    """
    base = _registry_dir(repo)
    by_hyp: Dict[str, Dict[str, Any]] = {}
    by_exp: Dict[str, Dict[str, Any]] = {}
    by_disp: Dict[str, Dict[str, Any]] = {}
    all_entries: List[Dict[str, Any]] = []
    sources: List[str] = []

    for filename in REGISTRY_FILES:
        path = base / filename
        if not path.exists():
            continue
        sources.append(filename)
        try:
            payload = json.loads(path.read_text(encoding="utf-8"))
        except Exception as exc:
            raise RuntimeError(f"failed to parse {path}: {exc}") from exc
        for entry in payload.get("hypotheses", []):
            hyp_id = entry.get("hypothesis_id")
            if not hyp_id:
                raise RuntimeError(f"{filename}: hypothesis missing hypothesis_id: {entry}")
            if hyp_id in by_hyp:
                raise RuntimeError(f"duplicate hypothesis_id: {hyp_id}")
            program = entry.get("program")
            if program not in VALID_PROGRAMS:
                raise RuntimeError(f"{hyp_id}: invalid program {program!r}")
            role = entry.get("role")
            if role not in VALID_ROLES:
                raise RuntimeError(f"{hyp_id}: invalid role {role!r}")
            by_hyp[hyp_id] = entry
            all_entries.append(entry)
            disp = entry.get("display_id")
            if disp:
                by_disp[disp] = entry
            for exp_id in entry.get("experiment_ids", []):
                if exp_id in by_exp:
                    raise RuntimeError(
                        f"experiment {exp_id} mapped to multiple hypotheses: "
                        f"{by_exp[exp_id].get('hypothesis_id')} and {hyp_id}"
                    )
                by_exp[exp_id] = entry

    base = {
        "schema_version": SCHEMA_VERSION,
        "by_hypothesis_id": by_hyp,
        "by_experiment_id": by_exp,
        "by_display_id": by_disp,
        "all": all_entries,
        "sources": sources,
    }
    if not apply_overlays:
        return base
    # Deferred import to avoid a cycle: hypothesis_proposals imports load_registry.
    from .hypothesis_proposals import apply_overlays_to_registry
    return apply_overlays_to_registry(base, repo)


def get_baseline_for_experiment(exp_id: str, repo: Path | str = ".") -> Optional[Dict[str, Any]]:
    return load_registry(repo)["by_experiment_id"].get(exp_id)


def list_baselines(repo: Path | str = ".") -> List[Dict[str, Any]]:
    return load_registry(repo)["all"]


def coverage_report(repo: Path | str = ".") -> Dict[str, Any]:
    """Verify every required experiment has a baseline hypothesis.

    Returns { covered: bool, missing: [...], extra: [...], total: int }.
    """
    reg = load_registry(repo)
    by_exp = reg["by_experiment_id"]
    missing = [exp for exp in REQUIRED_EXPERIMENT_IDS if exp not in by_exp]
    extra = [exp for exp in by_exp if exp not in REQUIRED_EXPERIMENT_IDS]
    return {
        "covered": not missing,
        "missing": missing,
        "extra": extra,
        "total": len(by_exp),
        "required": list(REQUIRED_EXPERIMENT_IDS),
    }


def baselines_by_role(repo: Path | str = ".") -> Dict[str, List[Dict[str, Any]]]:
    out: Dict[str, List[Dict[str, Any]]] = {role: [] for role in VALID_ROLES}
    for entry in load_registry(repo)["all"]:
        out.setdefault(entry["role"], []).append(entry)
    return out


def baselines_by_program(repo: Path | str = ".") -> Dict[str, List[Dict[str, Any]]]:
    out: Dict[str, List[Dict[str, Any]]] = {prog: [] for prog in VALID_PROGRAMS}
    for entry in load_registry(repo)["all"]:
        out.setdefault(entry["program"], []).append(entry)
    return out


def expected_baseline_status_values() -> Iterable[str]:
    return (
        "CONFIRMED",
        "FAILED",
        "INCOMPLETE",
        "INCONCLUSIVE",
        "NOT_APPLICABLE",
    )

File: lib/hypothesis-registry.ts
/**
 * Baseline hypothesis registry (TS side).
 *
 * Mirror of proof_kernel/hypothesis_registry.py. Reads the JSON registry files
 * under proof_kernel/hypotheses/ at runtime. Server-only (uses fs); never call
 * from browser bundles.
 *
 * The registry is read-only here. Agent-proposed revisions go through the
 * hypothesis_proposals workflow.
 */

import fs from "fs";
import path from "path";

export const HYPOTHESIS_REGISTRY_SCHEMA_VERSION = "2026.05.hypothesis-registry.v1";

export type HypothesisProgram = "PROGRAM_1" | "PROGRAM_2" | "NONE";

export type HypothesisRole =
    | "witness"
    | "control"
    | "pathfinder"
    | "demonstration"
    | "exploratory"
    | "visualization";

export type BaselineStatus =
    | "CONFIRMED"
    | "FAILED"
    | "INCOMPLETE"
    | "INCONCLUSIVE"
    | "NOT_APPLICABLE";

export interface ExpectedSignature {
    primary_metric: string;
    expected_value: string;
    tolerance: string;
    pass_rule: string;
}

export interface BaselineHypothesis {
    hypothesis_id: string;
    experiment_ids: string[];
    display_id: string;
    program: HypothesisProgram;
    role: HypothesisRole;
    plain_statement: string;
    object_under_test: string;
    expected_signature: ExpectedSignature;
    why_this_matters: string;
    failure_means: string[];
    failure_does_not_mean: string[];
    possible_alternative_hypotheses: string[];
    candidate_lemma_name?: string;
    intended_inference_if_passed?: string[];
    disallowed_conclusions?: string[];
}

export interface HypothesisRegistry {
    schema_version: string;
    byHypothesisId: Record<string, BaselineHypothesis>;
    byExperimentId: Record<string, BaselineHypothesis>;
    byDisplayId: Record<string, BaselineHypothesis>;
    all: BaselineHypothesis[];
    sources: string[];
}

export const REGISTRY_FILES = [
    "program_1.json",
    "program_2.json",
    "controls.json",
    "pathfinders.json",
    "demonstrations.json",
] as const;

export const REQUIRED_EXPERIMENT_IDS = [
    "EXP_0",
    "EXP_1",
    "EXP_1B",
    "EXP_1C",
    "EXP_2",
    "EXP_2B",
    "EXP_3",
    "EXP_4",
    "EXP_5",
    "EXP_6",
    "EXP_7",
    "EXP_8",
    "EXP_9",
    "EXP_10",
] as const;

const VALID_ROLES: ReadonlySet<HypothesisRole> = new Set([
    "witness",
    "control",
    "pathfinder",
    "demonstration",
    "exploratory",
    "visualization",
]);

const VALID_PROGRAMS: ReadonlySet<HypothesisProgram> = new Set([
    "PROGRAM_1",
    "PROGRAM_2",
    "NONE",
]);

function registryDir(repoRoot: string): string {
    return path.join(repoRoot, "proof_kernel", "hypotheses");
}

let cached: { repoRoot: string; overlayFingerprint: string; registry: HypothesisRegistry } | null = null;

const OVERLAY_FILENAME = "_accepted_overlays.json";

function overlayFingerprint(repoRoot: string): string {
    const fp = path.join(registryDir(repoRoot), OVERLAY_FILENAME);
    try {
        const stat = fs.statSync(fp);
        return `${stat.mtimeMs}:${stat.size}`;
    } catch {
        return "absent";
    }
}

interface OverlayEntry {
    proposal_id?: string;
    accepted_by?: string;
    accepted_at?: string;
    old_baseline_hash?: string;
    new_baseline_hash?: string;
    accepted_baseline?: Record<string, unknown>;
}

function readOverlayMap(repoRoot: string): Record<string, OverlayEntry> {
    const fp = path.join(registryDir(repoRoot), OVERLAY_FILENAME);
    if (!fs.existsSync(fp)) return {};
    try {
        const raw = JSON.parse(fs.readFileSync(fp, "utf-8")) as { overlays?: Record<string, OverlayEntry> };
        return raw.overlays ?? {};
    } catch {
        return {};
    }
}

function applyOverlays(
    base: HypothesisRegistry,
    repoRoot: string,
): HypothesisRegistry {
    const overlays = readOverlayMap(repoRoot);
    if (!Object.keys(overlays).length) return base;

    const byHyp: Record<string, BaselineHypothesis> = { ...base.byHypothesisId };
    const byExp: Record<string, BaselineHypothesis> = { ...base.byExperimentId };
    const byDisp: Record<string, BaselineHypothesis> = { ...base.byDisplayId };
    let all: BaselineHypothesis[] = [...base.all];

    for (const [hypId, ov] of Object.entries(overlays)) {
        const accepted = ov.accepted_baseline;
        if (!accepted || typeof accepted !== "object") continue;
        const previous = byHyp[hypId] ?? ({} as BaselineHypothesis);
        const merged = {
            ...previous,
            ...(accepted as Partial<BaselineHypothesis>),
            _overlay_provenance: {
                from_proposal_id: ov.proposal_id,
                accepted_by: ov.accepted_by,
                accepted_at: ov.accepted_at,
                old_baseline_hash: ov.old_baseline_hash,
                new_baseline_hash: ov.new_baseline_hash,
            },
        } as BaselineHypothesis & { _overlay_provenance: unknown };
        byHyp[hypId] = merged;
        for (const expId of merged.experiment_ids ?? []) byExp[expId] = merged;
        if (merged.display_id) byDisp[merged.display_id] = merged;
        const idx = all.findIndex((h) => h.hypothesis_id === hypId);
        if (idx >= 0) all[idx] = merged;
        else all = [...all, merged];
    }
    return {
        ...base,
        byHypothesisId: byHyp,
        byExperimentId: byExp,
        byDisplayId: byDisp,
        all,
    };
}

export function loadHypothesisRegistry(repoRoot: string = process.cwd()): HypothesisRegistry {
    const fingerprint = overlayFingerprint(repoRoot);
    if (cached && cached.repoRoot === repoRoot && cached.overlayFingerprint === fingerprint) {
        return cached.registry;
    }
    const dir = registryDir(repoRoot);
    const byHyp: Record<string, BaselineHypothesis> = {};
    const byExp: Record<string, BaselineHypothesis> = {};
    const byDisp: Record<string, BaselineHypothesis> = {};
    const all: BaselineHypothesis[] = [];
    const sources: string[] = [];

    for (const filename of REGISTRY_FILES) {
        const fp = path.join(dir, filename);
        if (!fs.existsSync(fp)) continue;
        sources.push(filename);
        let payload: { hypotheses?: BaselineHypothesis[] };
        try {
            payload = JSON.parse(fs.readFileSync(fp, "utf-8"));
        } catch (err) {
            throw new Error(`failed to parse ${filename}: ${(err as Error).message}`);
        }
        for (const entry of payload.hypotheses ?? []) {
            if (!entry.hypothesis_id) {
                throw new Error(`${filename}: hypothesis missing hypothesis_id`);
            }
            if (byHyp[entry.hypothesis_id]) {
                throw new Error(`duplicate hypothesis_id: ${entry.hypothesis_id}`);
            }
            if (!VALID_PROGRAMS.has(entry.program)) {
                throw new Error(`${entry.hypothesis_id}: invalid program ${entry.program}`);
            }
            if (!VALID_ROLES.has(entry.role)) {
                throw new Error(`${entry.hypothesis_id}: invalid role ${entry.role}`);
            }
            byHyp[entry.hypothesis_id] = entry;
            all.push(entry);
            if (entry.display_id) byDisp[entry.display_id] = entry;
            for (const expId of entry.experiment_ids ?? []) {
                if (byExp[expId]) {
                    throw new Error(
                        `experiment ${expId} mapped to multiple hypotheses: ${byExp[expId].hypothesis_id} and ${entry.hypothesis_id}`,
                    );
                }
                byExp[expId] = entry;
            }
        }
    }

    const baseRegistry: HypothesisRegistry = {
        schema_version: HYPOTHESIS_REGISTRY_SCHEMA_VERSION,
        byHypothesisId: byHyp,
        byExperimentId: byExp,
        byDisplayId: byDisp,
        all,
        sources,
    };
    const registry = applyOverlays(baseRegistry, repoRoot);
    cached = { repoRoot, overlayFingerprint: fingerprint, registry };
    return registry;
}

export function clearHypothesisRegistryCache(): void {
    cached = null;
}

export function getBaselineForExperiment(
    expId: string,
    repoRoot: string = process.cwd(),
): BaselineHypothesis | null {
    return loadHypothesisRegistry(repoRoot).byExperimentId[expId] ?? null;
}

export function getBaselineByHypothesisId(
    id: string,
    repoRoot: string = process.cwd(),
): BaselineHypothesis | null {
    return loadHypothesisRegistry(repoRoot).byHypothesisId[id] ?? null;
}

export function listBaselineHypotheses(
    repoRoot: string = process.cwd(),
): BaselineHypothesis[] {
    return loadHypothesisRegistry(repoRoot).all;
}

export interface HypothesisCoverageReport {
    covered: boolean;
    missing: string[];
    extra: string[];
    total: number;
    required: readonly string[];
}

export function hypothesisCoverageReport(
    repoRoot: string = process.cwd(),
): HypothesisCoverageReport {
    const reg = loadHypothesisRegistry(repoRoot);
    const present = new Set(Object.keys(reg.byExperimentId));
    const missing = REQUIRED_EXPERIMENT_IDS.filter((id) => !present.has(id));
    const extra = Array.from(present).filter(
        (id) => !(REQUIRED_EXPERIMENT_IDS as readonly string[]).includes(id),
    );
    return {
        covered: missing.length === 0,
        missing,
        extra,
        total: present.size,
        required: REQUIRED_EXPERIMENT_IDS,
    };
}
